src/full/Agda/TypeChecking/Unquote.hs

```diff
@@ -1041,6 +1044,22 @@ evalTCM v = do
       locallyReduceAllDefs $ checkFunDef NotDelayed i x cs
       primUnitUnit
 
+    tcPragmaForeign :: Text -> Text -> UnquoteM Term
+    tcPragmaForeign backend code = do
+      let x = [ForeignCode noRange $ T.unpack code]
+      -- FOREIGN pragmas may rely on those defined earlier so they are appended
+      lSnd . stForeignCode . key (T.unpack backend) %= Just . maybe x (++ x)
```

Comment:
Ah, I am not used to read lens code a lot.  The expression `key k %= Just . maybe x (++ x)` looks somewhat low-level to my eyes, and I cannot easily see what is going on.  Would you mind adding a helper function similiar to `addCompilerPragma` which I mention below?

Reply:
There is `addForeignCode : BackendName -> String -> TCM ()`. It prepends code though and uses `envRange` (which does I don't know what), should it be used?

Reply:
Looking at: https://github.com/agda/agda/blob/4b9617f778a328d19776423a1831143ae264f19f/src/full/Agda/TypeChecking/Monad/Base.hs#L197-L199
This uses a list but does not say whether you should append at the front or at the end.  Maybe this is an opportunity to clarify things.
Linked lists have bad performance when adding to the end, so I suppose the idea is to add a the front, and it is a stack.  The docs should really say that!  
This data then becomes part of the interface, from which the backend grabs it:
https://github.com/agda/agda/blob/4b9617f778a328d19776423a1831143ae264f19f/src/full/Agda/TypeChecking/Monad/Base.hs#L987
https://github.com/agda/agda/blob/4b9617f778a328d19776423a1831143ae264f19f/src/full/Agda/Compiler/MAlonzo/Pragmas.hs#L177-L182
I suppose because the order of declaration does not matter in a Haskell file, no attention is spent on whether the list is backwards of forwards here.

Two alternatives to proceed:
1. If you can live with it being a _stack_ (reversed list), maybe that is easiest, then please document it in the places I pointed out.  Then we add to the beginning, and the backend has to reverse the list if order matters.
2. If you want the list order being the same as the appearance order, and append at the end, then a Haskell list is not a good structure because of bad performance.  Please change it to e.g. `Data.Sequence` then and also document that one should add to the end.